### ISSUE TITLE
Feature request: implement a timeout when running formatter

### ISSUE THREAD

**Issue by ccoVeille** [hougesen/mdsf#139]
Your code launches formatter on each code snippet.

Based on what I can see the code is extracting the code snippet, look for the header, then call the formatter according to the one available

I was struggling with some files formating that were taking 700s to be formatted.

When I moved the file in /tmp/ it was instant.

I used strace to identify the issue. At first I thought the code was looking for tons of files because of a bug in discovering foldes.

It appears the code calls `npx` to launch formatter such as `sqlformatter`

npx relies on node_modules, so when the files were in a repository that needs a private npm registry, AND my VPN is down. It will try over and over to reach the npm registry that cannot be reached.

When I switched my VPN on, the mdsf was able to format instantly.

TL;DR; This should be considered as a source of problem. I think you should implement at timeout setting (that maybe configured) this was when formatting a file you could stop processing and inform a user there were an error.

**Comment by hougesen** [comment]
This makes a lot of sense. I'll look into it  🚀

**Comment by hougesen** [comment]
Okay, so I have been looking into this a bit.

I cannot see an easy way to implement a timeout, without having to refactor large parts of the tool.

Currently the way the formatter works is by creating a temporary file for each code block and then invoking each formatter sequentially on the file. In the end the file is read and the markdown file is updated.

By calling sigkill on an invoked formatter mid process we risk the formatter being stopped during a write operation, which could leave the code in a broken state.

The only way I can see to make a timeout work, is by converting the formatters to instead rely on stdin/stdout (Something like `cat $FILENAME | stylua -`), but that would add overhead since each codesnippet will be written to stdin/stdout multiple times. It also has the added problem of dealing with formatters that don't support stdin as input.

**Comment by ccoVeille** [comment]
Ouch indeed. I might be easy

What about a global timeout then? I mean not the lintee level and display error information.

But, it might be the exact way you thought about abd I didn't get it 😁

**Comment by ccoVeille** [comment]
Thanks for checking anyway

**Comment by hougesen** [comment]
A global timeout could work!

I’ll take a look at it once I have some spare time 😃

**Comment by ccoVeille** [comment]
Because my issue was your code npx so my VPN protected npm repository, but I wasn't on the VPN and without timeout I was waiting for nothing.

BTW, I think I saw you refactored and now local binary are tried before using an npx version

**Comment by hougesen** [comment]
> BTW, I think I saw you refactored and now local binary are tried before using an npx version

Yes, local binaries are ran before trying npx/bunx/deno in most cases.

I have been looking into a global timeout and it is actually blocked by the multithreading implementation in #330 since Rust does not support stopping/killing native threads 😅

I am going to benchmark the difference between using native ([std::thread](https://doc.rust-lang.org/std/thread/)) and green threads ([tokio::task](https://docs.rs/tokio/latest/tokio/task/index.html)) to see how big a difference it makes, if any.

I am currently expecting file io performance to be worse when using Tokio (async runtime for Rust) as described in [tokio-rs/tokio/issues/3664](https://github.com/tokio-rs/tokio/issues/3664).

I do not expect the performance of difference of command spawning to have any big impact since we are dealing with such a small amount of processes. I did find the finding of Jakub Beránek in [Process spawning performance in Rust](https://kobzol.github.io/rust/2024/01/28/process-spawning-performance-in-rust.html) rather interesting though.

**Comment by hougesen** [comment]
It looks like the difference between native and green threads is pretty negligible. Native threads did feel a lot smoother to look at, and impacted the overall performance of my computer a lot less. For that reason I am leaning towards not switching to green threads, and for that reason putting the timeout argument on hold.

I ran both on [rust-lang/rfcs](https://github.com/rust-lang/rfcs) which has a lot of markdown files.

![rust-lang/rfcs file breakdown](https://github.com/hougesen/mdsf/assets/56034786/054c06c7-9fa5-4e76-a93e-6575c4178b1e)

Left side is native, right is green threads.

- `mdsf format`

![mdsf format](https://github.com/hougesen/mdsf/assets/56034786/529f2259-7216-48d3-9ba5-55a0e28a9d66)

- `mdsf format --cache` with no cache

![mdsf format --cache without cache](https://github.com/hougesen/mdsf/assets/56034786/6eeb2028-b279-49f2-ab9b-0f9e04b6f8d3)

- `mdsf format --cache` with cache from previous operation

![mdsf format --cache with cache](https://github.com/hougesen/mdsf/assets/56034786/f14c4d37-863b-4f28-a39f-2a733cb0db87)

- `mdsf verify`

![mdsf verify](https://github.com/hougesen/mdsf/assets/56034786/e197b59f-14b1-49e1-9ebd-36b5898d8707)

**Comment by hougesen** [comment]
The async implementation can be tested here https://github.com/hougesen/mdsf/tree/refactor/async-runtime

This is also not a good sign for the async implementation 😅 

![image](https://github.com/hougesen/mdsf/assets/56034786/4945762e-9fc2-4f0e-9ea4-a1dd3061eea1)

**Comment by ccoVeille** [comment]
I'm impressed by the benchmark you made :heart:

**Comment by ccoVeille** [comment]
I find back this old discussion

Was it merged?

**Comment by hougesen** [comment]
No, sadly not.

I might take another crack at it, but for now I believe tasks like #389, #390 and #633 has a better reward/time-spent.

This task also has the added complication of practiacally requiring the tool to be rewritten to an async runtime (as mentioned above), which adds extra complexity.

**Comment by ccoVeille** [comment]
Even a global timeout would need a complete rewrite?

**Comment by hougesen** [comment]
Okay, so I have been playing around with it for the last few hours and I believe I have found a workaround that does not require async. 

I'll throw up a draft in a few moments.
